### PR TITLE
Fix parallel discovery

### PR DESCRIPTION
**Pull request by Evangelink**
## Description
The previously merged #3349 PR wasn't handling parallel discovery correctly for all cases.

## Related issue
Fixes [AB#1493062](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1493062)

Replaces #3435

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryEventsHandler.cs
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryEventsHandler.cs:67]:
```suggestion
        // We get DiscoveryComplete events from each ProxyDiscoveryManager (each testhost)
        // and they contain the last chunk of tests that were discovered in that particular testhost.
        // We want to send them to the IDE, but we don't want to tell it that discovery finished, because
        // other sources are still being discovered. So we translate the last chunk to a normal TestsDiscovered
        // event and send that to the IDE. Once all sources are completed, we will send a single DiscoveryComplete.
```

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryEventsHandler.cs
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryEventsHandler.cs:75]:
```suggestion
        // Do not send TestDiscoveryComplete to actual test discovery handler
        // we need to see if there are still sources left - let the parallel manager decide.
```

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryEventsHandler.cs
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryEventsHandler.cs:79]:
```suggestion
            lastChunk: null, // Set lastChunk to null, because we alreade sent the data using HandleDiscoveredTests above.
```

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryEventsHandler.cs
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryEventsHandler.cs:81]:
```suggestion

        if (!parallelDiscoveryComplete)
```

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs:243]: I don't like this nullcheck, if there were no extra sources to be discovered and we did not abort, then it would be logical to send null. Is this just a substitute for checking IsAborted?

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/Parallel/ParallelDiscoveryDataAggregatorTests.cs
Comment (nohwnd) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/Parallel/ParallelDiscoveryDataAggregatorTests.cs:23]: I might be too tired today, but what is the change in syntax? is that going from method call with parameters to, method call that takes a single object, which has non-default constructor?

Comment (Evangelink) [reply test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/Parallel/ParallelDiscoveryDataAggregatorTests.cs:23]: The aggregate used to take 2 args (total tests and is aborted), now it takes the event arg (that has an overload with the previous 2 args).

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryEventsHandler.cs
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryEventsHandler.cs:81]:
```suggestion
            lastChunk: null, // Set lastChunk to null, because we already sent the data using HandleDiscoveredTests above.
```

Comment (nohwnd) [reply src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryEventsHandler.cs:81]: Sorry, this was my typo :/

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs
Comment (Evangelink) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs:240]: From the various tests I made with the playground project, I haven't found any case where this would be useful. I don't know if that's ok to remove or if there is a particular case I missed.

File: src/Microsoft.TestPlatform.ObjectModel/Client/Events/DiscoveryCompleteEventArgs.cs
Comment (Evangelink) [inline src/Microsoft.TestPlatform.ObjectModel/Client/Events/DiscoveryCompleteEventArgs.cs:30]: I had to remove this comment because we are now also using this event args inside the parallel discovery.

**Comment by Evangelink** [comment]
I still need to make some manual tests before we merge this PR.

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs
Comment (Evangelink) [inline test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs:217]: Doesn't change much for the discovery cancelled case but without this setup the tested condition was always true even when not cancelling.

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs
Comment (Evangelink) [inline test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs:206]:
```suggestion
            GetAssetFullPath("DiscoveryTestProject.dll"),
            GetAssetFullPath("SimpleTestProject.dll"),
```

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs
Comment (nohwnd) [inline test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs:230]: This seems flaky, does this need to cancel on some particular test? Maybe you can cancel after getting first batch of tests / first batch of tests from a given source that has many of them?

Comment (Evangelink) [reply test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs:230]: There is no specific needs but putting the cancellation after the discovery of some sources makes the test really slow locally (around 30s) for the net451 TFM

image

Comment (Evangelink) [reply test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs]: I have updated the test to have a better logic and prevent flakiness.

File: test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs
Comment (Evangelink) [reply test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/DiscoverTests.cs:222]: Comment is no longer relevant as we now check that we correctly signal abortion.

File: test/Microsoft.TestPlatform.AcceptanceTests/AcceptanceTestBase.cs
Comment (Evangelink) [inline test/Microsoft.TestPlatform.AcceptanceTests/AcceptanceTestBase.cs:114]: Unrelated to this PR but fixes the failing acceptance tests that were blocking this PR.

File: test/Microsoft.TestPlatform.AcceptanceTests/DotnetTestTests.cs
Comment (Evangelink) [inline test/Microsoft.TestPlatform.AcceptanceTests/DotnetTestTests.cs:4]: Changes to this file are unrelated to this PR but fixes the failing acceptance tests that were blocking this PR.

File: test/Microsoft.TestPlatform.AcceptanceTests/AcceptanceTestBase.cs
Comment (Evangelink) [inline test/Microsoft.TestPlatform.AcceptanceTests/AcceptanceTestBase.cs:5]: Changes to this file are unrelated to this PR but fixes the failing acceptance tests that were blocking this PR.

File: test/Microsoft.TestPlatform.AcceptanceTests/ResultsDirectoryTests.cs
Comment (Evangelink) [inline test/Microsoft.TestPlatform.AcceptanceTests/ResultsDirectoryTests.cs:4]: Changes to this file are unrelated to this PR but fixes the failing acceptance tests that were blocking this PR.

File: test/Microsoft.TestPlatform.AcceptanceTests/RunsettingsTests.cs
Comment (Evangelink) [inline test/Microsoft.TestPlatform.AcceptanceTests/RunsettingsTests.cs:11]: Changes to this file are unrelated to this PR but fixes the failing acceptance tests that were blocking this PR.

File: playground/TestPlatform.Playground/Program.cs
Comment (Evangelink) [inline playground/TestPlatform.Playground/Program.cs:66]: Kept it so that we don't have to create it again in other playground runs.

File: src/Microsoft.TestPlatform.Common/Microsoft.TestPlatform.Common.csproj
Comment (Evangelink) [inline src/Microsoft.TestPlatform.Common/Microsoft.TestPlatform.Common.csproj:9]: Adding net6.0 because I need a TFM greater or equal to netcoreapp3.0 so that we have a dll where `NullableAttributes.cs` is not included. Otherwise we get an error that says the attribute comes from 2 locations (system and this). I am using net6.0 because we are already targeting this for `DotNetBuildFromSource`.

File: src/Microsoft.TestPlatform.Common/Utilities/Debug.cs
Comment (Evangelink) [inline src/Microsoft.TestPlatform.Common/Utilities/Debug.cs:12]: I am always forcing this one but FYI in recent versions the system `Debug.Assert` methods already support it.

File: src/Microsoft.TestPlatform.ObjectModel/Client/Events/DiscoveryCompleteEventArgs.cs
Comment (Evangelink) [inline src/Microsoft.TestPlatform.ObjectModel/Client/Events/DiscoveryCompleteEventArgs.cs:62]: I had to remove this comment because we are now also using this event args inside the parallel discovery.

File: src/Microsoft.TestPlatform.ObjectModel/Client/Events/DiscoveryCompleteEventArgs.cs
Comment (Evangelink) [inline src/Microsoft.TestPlatform.ObjectModel/Client/Events/DiscoveryCompleteEventArgs.cs:27]: @cvpoienaru added a new ctor with this signature + the extensions dictionary but this public ctor is not yet released so we can just go with only one overload.

File: playground/TestPlatform.Playground/Program.cs
Comment (nohwnd) [reply playground/TestPlatform.Playground/Program.cs:66]: Not necessary in this PR but we should make the handler do something similar as the other handler does, to make it more useful.

File: src/Microsoft.TestPlatform.Common/Microsoft.TestPlatform.Common.csproj
Comment (nohwnd) [reply src/Microsoft.TestPlatform.Common/Microsoft.TestPlatform.Common.csproj:9]: Just to confirm. This is not being shipped in any of our Nugets?

File: src/Microsoft.TestPlatform.Common/Utilities/Debug.cs
Comment (nohwnd) [reply src/Microsoft.TestPlatform.Common/Utilities/Debug.cs:11]: Keep it.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryDataAggregator.cs
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryDataAggregator.cs:86]: I doubt anyone is using this "mandated" value, but let's keep it just in case.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryDataAggregator.cs
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryDataAggregator.cs:98]:
```suggestion
        // because this aggregator is shared with proxies so the state is already up-to-date.
        //
        // The reason to share the aggregator is that we only get 1 notification for completed discovery
        // when discovery is done Instead of 1 notification per discovery manager. And we also don't get 
        // any notification for discovery of sources that did not start discovering yet. So we need the state
        // in the aggregator to start with all the sources that we were requested to discover in NotDiscovered
        // state. And then as we discover them, we need to update that directly. So when any discovery is cancelled
        // we can just take the current state in the aggregator and report it.
```

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryDataAggregator.cs
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryDataAggregator.cs:196]: Is this still true, is this code now sharing all sources in the "cache"? And do we trigger the isComplete from parallel manager or each proxy manager? If we do it from each, don't we mark every source as discovered for every proxy manager that was able to finish?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs:254]: this is per proxy, and marks it as complete, which I think we make all sources complete, even the ones we don't own.

File: playground/TestPlatform.Playground/Program.cs
Comment (Evangelink) [reply playground/TestPlatform.Playground/Program.cs:66]: Done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryDataAggregator.cs
Comment (Evangelink) [reply src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryDataAggregator.cs:86]: Text was just moved from another part :)

File: src/Microsoft.TestPlatform.Common/Microsoft.TestPlatform.Common.csproj
Comment (Evangelink) [reply src/Microsoft.TestPlatform.Common/Microsoft.TestPlatform.Common.csproj:9]: I have just double checked and running build.cmd locally doesn't produce the net6.0 of the dll in the artifacts folder.

**Review by nohwnd — APPROVED** [review]

**Pull request merged**
